examples/vision/metric_learning.py

```diff
@@ -22,290 +24,380 @@
-y_test = np.squeeze(y_test)
+print('TensorFlow:', tf.__version__)
+print('TensorFlow Similarity', tfsim.__version__)
 
 """
-To get a sense of the dataset we can visualise a grid of 25 random examples.
-
-
+# Dataset Samplers
+
+We will be using the
+[CIFAR-10](https://www.tensorflow.org/datasets/catalog/cifar10) dataset for this
+tutorial.
+
+For a similarity model to learn efficiently, each batch must contains at least 2
+examples of each class.
+
+To make this easy, tf_similarity offers `Samplers()` that enable you to set both
+the number of classes and the minimum number of examples of each class per
+batch.
+
+The train and test datasets will be created using the
+`TFDatasetMultiShotMemorySampler()`. This creates a sampler that loads datasets
+from [tensorflow datasets](https://www.tensorflow.org/datasets) and yields
+batches containing a target number of classes and a target number of examples
+per class. Additionally, we can restrict the sampler to only yield the subset of
+classes defined in class_list, enabling us to train on a subset of the classes
+and then test how the embedding generalizes to the unseen classes. This can be
+useful when working on few-shot learning problems.
+
+The following cell creates a train_ds sample that:
+
+* Loads the cifar10 dataset from tfds and then takes the
+examples_per_class_per_batch.
+* Normalizes the images by applying the preprocess function to each image.
+* Ensures the sampler restricts the classes to those defined in the class_list.
+* Ensures each batch contains 4 different classes with 16 examples each.
+
+We also create a train_ds in the same way, but we limit the total number of
+examples per class to 100 and the examples per class per batch is set to the
+default of 2.
 """
 
-height_width = 32
 
+def normalize(img, label):
+    "Scale the images to between [0,1]."
+    img = tf.cast(img, dtype='float32') / 255.0
+    return img, label
 
-def show_collage(examples):
-    box_size = height_width + 2
-    num_rows, num_cols = examples.shape[:2]
 
-    collage = Image.new(
-        mode="RGB",
-        size=(num_cols * box_size, num_rows * box_size),
-        color=(250, 250, 250),
-    )
-    for row_idx in range(num_rows):
-        for col_idx in range(num_cols):
-            array = (np.array(examples[row_idx, col_idx]) * 255).astype(np.uint8)
-            collage.paste(
-                Image.fromarray(array), (col_idx * box_size, row_idx * box_size)
-            )
+examples_per_class_per_batch = 8
+classes_per_batch = 10
+class_list = random.sample(range(10), 10)
 
-    # Double size for visualisation.
-    collage = collage.resize((2 * num_cols * box_size, 2 * num_rows * box_size))
-    return collage
+print(" Create Training Data ".center(34, '#'))
+train_ds = TFDatasetMultiShotMemorySampler(
+    'cifar10',
+    classes_per_batch=classes_per_batch,
+    splits='train',
+    steps_per_epoch=5000,
+    examples_per_class_per_batch=examples_per_class_per_batch,
```

Comment:
discuss why you want 8 or even more per class --> I think if you are using a TPU you need way more to saturate it.

Reply:
Done